### PR TITLE
Add StateMetadataFeature

### LINKED ISSUE #1650: Add StateMetadataFeature
https://github.com/iotaledger/tips/blob/tip54/tips/TIP-0054/tip-0054.md#state-metadata-feature

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Add StateMetadataFeature

## Links to any relevant issues

Closes #1650 

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

cargo test

### CODE REVIEW

File: sdk/src/types/block/output/feature/state_metadata.rs
Comment (Alex6323) [inline sdk/src/types/block/output/feature/state_metadata.rs:1]: Don't know if we want that but we could make the packable `verify..` methods generic over `T` and move them to `mod.rs` to re-use them?

Comment (Thoralf-M) [reply sdk/src/types/block/output/feature/state_metadata.rs:1]: I also thought about it, but what would `T` be? Is it even possible to have something generic that can be used like this then `T::BYTE_LENGTH_RANGE` (in verify_byte_length_packable())?

Comment (DaughterOfMars) [reply sdk/src/types/block/output/feature/state_metadata.rs:1]: tbh we already have lots of duplicated code, like between output types. I feel like we shouldn't stress too much about adding a little more for this

Comment (Alex6323) [reply sdk/src/types/block/output/feature/state_metadata.rs:1]: yeah, let's not get hung up on that

**Review by Alex6323 — DISMISSED** [review]

File: sdk/src/types/block/output/anchor.rs [hunk 1]
```diff
@@ -373,13 +373,6 @@ impl AnchorOutput {
         self.state_index
     }
 
-    // TODO https://github.com/iotaledger/iota-sdk/issues/1650
-    // ///
-    // #[inline(always)]
-    // pub fn state_metadata(&self) -> &[u8] {
```
Comment (thibault-martinez) [inline sdk/src/types/block/output/anchor.rs:379]: I think a getter would still be useful tbh

File: sdk/src/types/block/output/feature/mod.rs [hunk 2]
```diff
@@ -55,6 +57,9 @@ pub enum Feature {
     /// A metadata feature.
     #[packable(tag = MetadataFeature::KIND)]
     Metadata(MetadataFeature),
+    /// A state metadata feature.
+    #[packable(tag = StateMetadataFeature::KIND)]
+    StateMetadata(StateMetadataFeature),
```
Comment (thibault-martinez) [inline sdk/src/types/block/output/feature/mod.rs:62]: You need to bump the type bytes of the features after this one

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: sdk/src/types/block/output/feature/metadata.rs [hunk 3]
```diff
@@ -25,13 +25,13 @@ pub(crate) type MetadataFeatureEntryCount = BoundedU8<1, { u8::MAX }>;
 pub(crate) type MetadataFeatureKeyLength = BoundedU8<1, { u8::MAX }>;
 pub(crate) type MetadataFeatureValueLength = BoundedU16<0, { u16::MAX }>;
 
-type MetadataBTreeMapPrefix = BTreeMapPrefix<
+pub(crate) type MetadataBTreeMapPrefix = BTreeMapPrefix<
     BoxedSlicePrefix<u8, MetadataFeatureKeyLength>,
     BoxedSlicePrefix<u8, MetadataFeatureValueLength>,
     MetadataFeatureEntryCount,
 >;
```
Comment (thibault-martinez) [inline sdk/src/types/block/output/feature/metadata.rs]:
```suggestion
```

File: sdk/src/types/block/output/feature/mod.rs (hunk 2 above)
Comment (thibault-martinez) [reply sdk/src/types/block/output/feature/mod.rs:62]: Oh that was already done, ignore me

File: sdk/src/types/block/output/anchor.rs (hunk 1 above)
Comment (thibault-martinez) [reply sdk/src/types/block/output/anchor.rs:379]: But we don't really do it for the others so let's forget

File: sdk/src/types/block/output/feature/state_metadata.rs [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+                })
+                .collect::<BTreeMap<_, _>>();
+
+            Self {
+                kind: StateMetadataFeature::KIND,
+                entries,
+            }
+        }
+    }
+    impl Serialize for StateMetadataFeature {
```
Comment (thibault-martinez) [inline sdk/src/types/block/output/feature/state_metadata.rs:284]:
```suggestion

    impl Serialize for StateMetadataFeature {
```

**Review by thibault-martinez — APPROVED** [review]

**Review by DaughterOfMars — APPROVED** [review]

**Pull request merged**